src/ppx/xr_component.cpp

```diff
@@ -0,0 +1,395 @@
+    PPX_ASSERT_MSG(swapchains.size() == viewCount, "Swapchain needs to match number of views!");
+    for (uint32_t i = 0; i < viewCount; ++i) {
+        mCompositionLayerProjectionViews[i]                           = {XR_TYPE_COMPOSITION_LAYER_PROJECTION_VIEW};
+        mCompositionLayerProjectionViews[i].pose                      = mViews[i].pose;
+        mCompositionLayerProjectionViews[i].fov                       = mViews[i].fov;
+        mCompositionLayerProjectionViews[i].subImage.swapchain        = swapchains[i]->GetXrSwapchain();
+        mCompositionLayerProjectionViews[i].subImage.imageRect.offset = {0, 0};
+        mCompositionLayerProjectionViews[i].subImage.imageRect.extent = {static_cast<int>(GetWidth()), static_cast<int>(GetHeight())};
+    }
+
+    // Begin frame
+    XrFrameBeginInfo frameBeginInfo = {
+        .type = XR_TYPE_FRAME_BEGIN_INFO,
+    };
+
+    XrResult result = xrBeginFrame(mSession, &frameBeginInfo);
+    if (result != XR_SUCCESS) {
+        switch (result) {
+            case XR_SESSION_LOSS_PENDING:
+            case XR_FRAME_DISCARDED:
+                mShouldRender = false;
+                break;
+
+            default:
+                PPX_ASSERT_MSG(false, "xrBeginFrame failed!");
+                break;
+        }
+    }
+}
+
+void XrComponent::EndFrame()
+{
+    mCompositionLayerProjection.space     = mRefSpace;
+    mCompositionLayerProjection.viewCount = static_cast<uint32_t>(GetViewCount());
+    mCompositionLayerProjection.views     = mCompositionLayerProjectionViews.data();
+    XrCompositionLayerBaseHeader* layer   = mShouldRender ? reinterpret_cast<XrCompositionLayerBaseHeader*>(&mCompositionLayerProjection) : nullptr;
+
+    XrFrameEndInfo frameEndInfo = {
+        .type                 = XR_TYPE_FRAME_END_INFO,
+        .displayTime          = mFrameState.predictedDisplayTime,
+        .environmentBlendMode = mBlend,
+        .layerCount           = static_cast<uint32_t>(mShouldRender ? 1 : 0),
+        .layers               = &layer,
+    };
+
+    VERIFY_XR_CALL(xrEndFrame(mSession, &frameEndInfo));
+}
+
+glm::mat4 XrComponent::GetCurrentViewViewMatrix() const
+{
+    PPX_ASSERT_MSG((mCurrentViewIndex < mViews.size()), "Invalid view index!");
+    const XrView&  view = mViews[mCurrentViewIndex];
+    const XrPosef& pose = view.pose;
+    // OpenXR is using right handed system which is the same as Vulkan
+    glm::quat quat         = glm::quat(pose.orientation.w, pose.orientation.x, pose.orientation.y, pose.orientation.z);
+    glm::mat4 rotation     = glm::mat4_cast(quat);
+    glm::vec3 position     = glm::vec3(pose.position.x, pose.position.y, pose.position.z);
+    glm::mat4 translation  = glm::translate(glm::mat4(1.f), position);
+    glm::mat4 view_glm     = translation * rotation;
+    glm::mat4 view_glm_inv = glm::inverse(view_glm);
+    return view_glm_inv;
+}
+
+glm::mat4 XrComponent::GetCurrentViewProjectionMatrix(float nearZ, float farZ) const
```

Comment:
This is just the projection matrix, isn't it?

Reply:
it is "Get current view's projection matrix". maybe it is better to call it GetProjectionMatrixOfCurrentView?

Reply:
Yep, sounds better!